amethyst_network/src/error.rs

```diff
@@ -0,0 +1,78 @@
+//! Module containing error handling logic.
+
+use std::{
+    fmt::{self, Display, Formatter},
+    io,
+    sync::mpsc,
+};
+
+use crate::server::ServerSocketEvent;
+
+/// The `amethyst_network` result type.
+pub type Result<T> = std::result::Result<T, ErrorKind>;
+
+/// Wrapper for all errors who could occur in `amethyst_network`.
+#[derive(Debug)]
+pub enum ErrorKind {
+    /// Error that could occur on the UDP-socket
+    UdpError(laminar::error::NetworkError),
+    /// Error that could occur whit IO.
+    IoError(io::Error),
+    /// Error that could occur when serializing whit `bincode`
+    SerializeError(bincode::ErrorKind),
+    /// Error that could occur when sending an `ServerSocketEvent` to some channel.
+    ChannelSendError(mpsc::SendError<ServerSocketEvent>),
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+
+impl std::error::Error for ErrorKind {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(self)
+    }
```

Comment:
Right, that means that you should remove `cause` completely and _only_ implement `source`.

The default implementation of `source` is returning `None`: https://doc.rust-lang.org/src/std/error.rs.html#206
The default implementation of `cause` calls `source`: https://doc.rust-lang.org/src/std/error.rs.html#143

Reply:
This is not fixed yet afaics?

Reply:
😑 I'm totally confused now, I'll take another try. I think I am missing interpreting  your suggestions. Do you have an code exaple?

Reply:
And is this really needed to change, since when the new error handling crate from your PR is merged all this error handling code will probably change.

Reply:
This is what it should look like (it should look familiar since it's what you initially did in `cause`):
```rust
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match *self {
            ErrorKind::UdpError(ref e) => Some(e),
            ErrorKind::IoError(ref e) => Some(e),
            ErrorKind::ChannelSendError(ref e) => Some(e),
            ErrorKind::SerializeError(ref e) => Some(e),
            _ => None,
        }
    }
```

This is necessary because you are implementing a custom error type. Not using the dynamic one introduced in #1220.

The reason why your approach would cause an infinite loop is because you walk through the error chain like this:

```rust
let mut e = ...;

println!("Error: {}", e);

while let Some(s) = e.source() {
    println!("Caused by: {}", s);
    e = s;
}
```

And if `source` returns `Some(self)` this is gonna loop forever.

Check out the documentation for source for a more complete example on this:
https://doc.rust-lang.org/std/error/trait.Error.html#method.source

Reply:
Thanks for writing that up!!. I'll see what I can do whit this.